client-sdk/rust/src/client.rs

```diff
@@ -0,0 +1,374 @@
+use std::{marker::PhantomData, sync::Arc};
+
+use bytes::{Buf as _, BufMut as _};
+use futures_util::{
+    future::try_join_all,
+    stream::{Stream, TryStreamExt},
+};
+use prost::Message as _;
+use tonic::{self, client::Grpc, transport::Channel};
+
+use oasis_runtime_sdk::{
+    self as sdk,
+    core::{
+        common::{crypto::hash::Hash, namespace::Namespace},
+        consensus::roothash::{AnnotatedBlock, Block},
+        transaction::tags::Tag,
+    },
+    types::transaction::{
+        AuthInfo, Call, CallFormat, Fee, SignerInfo, Transaction, LATEST_TRANSACTION_VERSION,
+    },
+};
+
+use crate::{requests::*, types::Round, wallet::Wallet};
+
+/// A sentinel value for the latest round.
+const ROUND_LATEST: u64 = u64::max_value();
+
+/// The prefix of the runtime chain context including the signature context,
+/// `oasis-runtime-sdk/tx: v1` followed by the separator, ` for chain `.
+const CHAIN_CONTEXT_PREFIX: &str = "oasis-runtime-sdk/tx: v0 for chain";
+
+#[derive(Clone)]
+pub struct Client {
+    inner: Grpc<Channel>, // Cheap to `Clone`, so no `Arc`
+    runtime_id: Namespace,
+    wallets: Arc<Vec<Arc<dyn Wallet>>>,
+    fee: Fee,
+    chain_context: Vec<u8>,
+}
+
+impl Client {
+    /// Connects to the oasis-node listening on Unix socket at `sock_path` communicating
+    /// with the identified runtime. Transactions will be signed by the `signer`.
+    /// Do remember to call `set_fee` as appropriate before making the first call.
+    pub async fn connect(
+        sock_path: impl AsRef<std::path::Path> + Clone + Send + Sync + 'static,
+        runtime_id: Namespace,
+        wallets: impl IntoIterator<Item = Box<dyn Wallet>>,
+    ) -> Result<Self, Error> {
+        let channel = tonic::transport::Channel::from_static(
+            "://.", /* Unused, but required to be a URI. */
+        )
+        .connect_with_connector(tower::service_fn(move |_| {
+            tokio::net::UnixStream::connect(sock_path.clone())
+        }))
+        .await?;
+        Self::connect_through_channel(channel, runtime_id, wallets).await
+    }
+
+    pub async fn connect_through_channel(
+        channel: tonic::transport::Channel,
+        runtime_id: Namespace,
+        wallets: impl IntoIterator<Item = Box<dyn Wallet>>,
+    ) -> Result<Self, Error> {
+        let mut grpc = Grpc::new(channel);
+
+        let consensus_chain_context =
+            Self::make_unary(&mut grpc, GetChainContextRequest {}).await?;
+        let runtime_chain_context =
+            Hash::digest_bytes_list(&[&runtime_id.0, &consensus_chain_context]);
+        let chain_context = format!("{} {:x}", CHAIN_CONTEXT_PREFIX, runtime_chain_context);
+
+        Ok(Self {
+            inner: grpc,
+            runtime_id,
+            wallets: Arc::new(wallets.into_iter().map(Arc::from).collect()),
+            fee: Default::default(),
+            chain_context: chain_context.into_bytes(),
+        })
+    }
+
+    pub fn set_fee(&mut self, fee: Fee) {
```

Comment:
it's not global. it's per-client. I don't know how fees will work in practice, so this isn't a very full-featured implementation.

Reply:
By global I meant across all transactions generated by the client. It would make more sense for this to be part of a transaction builder (see Go client).